src/pubsub/ua_pubsub_reader.c

```diff
@@ -1218,6 +1207,58 @@ UA_Server_DataSetReader_addTargetVariables(UA_Server *server, UA_NodeId *parentN
     return retval;
 }
 
+void UA_Server_DataSetReader_processRTMessage(UA_Server *server, UA_DataSetReader *dataSetReader) {
+    /* Check the decoded message is the expected one
+    * TODO: PublisherID check after modification in NM to support all datatypes */
+    if((dataSetReader->bufferedMessage.nm->groupHeader.writerGroupId != dataSetReader->config.writerGroupId) ||
+       (*dataSetReader->bufferedMessage.nm->payloadHeader.dataSetPayloadHeader.dataSetWriterIds != dataSetReader->config.dataSetWriterId)) {
+        UA_LOG_INFO(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                    "PubSub receive. Unknown message received. Will not be processed.");
+        return;
+    }
+
+    UA_DataSetMessage* dataSetMsg = dataSetReader->bufferedMessage.nm->payload.dataSetPayload.dataSetMessages;
+    if(!dataSetMsg->header.dataSetMessageValid) {
+        UA_LOG_INFO(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                    "DataSetMessage is discarded: message is not valid");
+        return;
+    }
+
+    if(dataSetMsg->header.dataSetMessageType == UA_DATASETMESSAGE_DATAKEYFRAME) {
+        if(dataSetMsg->header.fieldEncoding != UA_FIELDENCODING_RAWDATA) {
+            size_t anzFields = dataSetMsg->data.keyFrameData.fieldCount;
+            if(dataSetReader->subscribedDataSetTarget.targetVariablesSize < anzFields)
+                anzFields = dataSetReader->subscribedDataSetTarget.targetVariablesSize;
+
+            for(UA_UInt16 i = 0; i < anzFields; i++) {
+                if(dataSetMsg->data.keyFrameData.dataSetFields[i].hasValue) {
+                    const UA_VariableNode *rtNode = (const UA_VariableNode *)
+                                                     UA_NODESTORE_GET(server, &dataSetReader->subscribedDataSetTarget.targetVariables[i].targetNodeId);
```

Comment:
Calling `UA_NODESTORE_GET` is where you loose real time guarantees.
Because this requires synchronisation with the main server.

Instead, move the stuff from the valueBackend directly to the PubSub config.
So you don't have to do the lookup in every cycle.

Reply:
Need some clarifications here. 
I can understand that new struct member UA_DataValue should be created in DSR config so we can prevent using UA_NODESTORE_GET in ua_pubsub_reader.c while receiving. But I could not create UA_DataValue for each fields as there is no DataSetField config in subscriber (it uses fieldmetadata in datasetmetadata) like it is there in publisher.

For this, I have mapped the values directly into the KeyValuePair of FieldMetaData and used this in ua_pubsub_reader.c while receiving. Does this hold good?